third_party/move/move-vm/runtime/src/storage/ty_tag_converter.rs

```diff
@@ -12,6 +12,56 @@ use move_vm_types::loaded_data::{runtime_types::Type, struct_name_indexing::Stru
 use parking_lot::RwLock;
 use std::hash::{Hash, Hasher};
 
+struct PseudoGasContext {
+    // Parameters for metering type tag construction:
+    //   - maximum allowed cost,
+    //   - base cost for any type to tag conversion,
+    //   - cost for size of a struct tag.
+    max_cost: u64,
+    cost: u64,
+    cost_base: u64,
+    cost_per_byte: u64,
+}
+
+impl PseudoGasContext {
+    fn new(vm_config: &VMConfig) -> Self {
+        Self {
+            max_cost: vm_config.type_max_cost,
+            cost: 0,
+            cost_base: vm_config.type_base_cost,
+            cost_per_byte: vm_config.type_byte_cost,
+        }
+    }
+
+    fn current_cost(&mut self) -> u64 {
+        self.cost
+    }
+
+    fn charge_base(&mut self) -> PartialVMResult<()> {
+        self.charge(self.cost_base)
+    }
+
+    fn charge_struct_tag(&mut self, struct_tag: &StructTag) -> PartialVMResult<()> {
+        let size =
+            (struct_tag.address.len() + struct_tag.module.len() + struct_tag.name.len()) as u64;
+        self.charge(size * self.cost_per_byte)
+    }
+
+    fn charge(&mut self, amount: u64) -> PartialVMResult<()> {
+        self.cost += amount;
+        if self.cost > self.max_cost {
+            Err(
+                PartialVMError::new(StatusCode::TYPE_TAG_LIMIT_EXCEEDED).with_message(format!(
```

Comment:
super corner case when it could be useful (e.g. there was a bug that charged some crazy amount), but might as well add the amount to the message, i.e.
After charging {amount} exceeded ..